src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs

```diff
@@ -485,6 +486,13 @@ private async Task HandleChallengeAsyncInternal(AuthenticationProperties propert
                 "Cannot redirect to the authorization endpoint, the configuration may be missing or invalid.");
         }
 
+        var parRequired = ConfigFlagEnabled("require_pushed_authoriation_requests");
+
+        if (Options.UsePushedAuthorization || parRequired)
```

Comment:
Do we need a way for the developer to say use pushed authorization if possible, but don't fail if there's no "pushed_authorization_request_endpoint" available?

Reply:
Do we expect this support to be variable? E.g. an auth server will support it or not, you should be able to determine that during development. If that support changes later I don't know that we should react dynamically in prod to it as there may be side-effects.

Reply:
The change in behavior seems small. It doesn't even seem to prevent any callbacks from being called. It does change the parameters used for the user agent's ultimate authorization request, but I don't see how that's likely to cause issues. I'd be more worried about the OIDC provider requiring a custom authentication scheme to use the PAR endpoint, but we could fall back to not using PAR if given a non-200 response if PAR wasn't explicitly opted into. We could even stop trying PAR if the first attempt fails.

I think it's very possible that an OIDC provider that didn't support PAR during initial development might add support later. I understand not wanting to risk any potential breaks caused by an OIDC provider lighting up a feature that's not actually required for us to use, but it feels like we should default to the more secure option if it's available and it's unlikely to break anyone.

@blowdart do you have an opinion on this?

Reply:
An apps behavior should be in control of the developer, not a 3rd party, and shouldn't change without the developer instigating and testing that change.

Reply:
The `OpenIdConnectHandler` lighting up features only if the `.well-known/openid-configuration` indicates the OIDC provider supports it is not breaking new ground. If you set `GetClaimsFromUserInfoEndpoint` to true, we will ignore it (aside from logging a debug-level message) if there is no "userinfo_endpoint".

https://github.com/dotnet/aspnetcore/blob/53db4d97d7c77d13e20e58a98f104e88d6af6040/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L934-L940

We can debate what the default should be, but I think it's reasonable to want to have a way to say if the OIDC provider supports a feature that should enhance security, use it. I could see somebody wanting to do this in a library or code analyzer that ensures PAR is used when available.

Reply:
One thing that I've encountered is that some providers actually support PAR, but not for all clients. We ran into this when we added similar logic to IdentityModel.OidcClient (https://github.com/IdentityModel/IdentityModel.OidcClient/issues/428).